### PR TITLE
Add `prefer-true-attribute-shorthand` rule

### LINKED ISSUE #775: Rule Proposal: Boolean prop shorthand (`prop` instead of `:prop="true"`)
**Please describe what the rule should do:**

I've seen code like `<div :boolProp="true" />` often where you could use `<div boolProp />` instead. I'd love to see a rule for that (plus an auto-fix option which should be fairly possible)

**What category should the rule belong to?**
- [x] Enforces code style
- [ ] Warns about a potential error
- [x] Suggests an alternate way of doing something
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**
```vue
<template>
  <div>
    <button :disabled="true">
  </div>
</template>
```

### LINKED ISSUE #1781: New rule: prefer-true-attribute-shorthand
**Please describe what the rule should do:**

In template, if there's an attribute whose value is `true`, such as `:show="true"`, this rule will warn and provide auto fix (becomes `show`, see the example below). This can reduce verbosity.

**What category should the rule belong to?**

[x] Suggests an alternate way of doing something (suggestion)

**Provide 2-3 code examples that this rule should warn about:**

Invalid:

```vue
<template>
  <my-dialog :show="true" />
</template>
```

which can be fixed to:

```vue
<template>
  <my-dialog show />
</template>
```

### PR DESCRIPTION
**Pull request by g-plane**
Closes #775 
Closes #1781

### CODE REVIEW

**Review by ota-meshi — APPROVED** [review]
LGTM! Thank you!

File: lib/rules/prefer-true-attribute-shorthand.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'require shorthand form attribute when `v-bind` value is `true`.',
```
Comment (FloEdelmann) [inline lib/rules/prefer-true-attribute-shorthand.js:22]: To stay consistent with other descriptions:

```suggestion
        'require shorthand form attribute when `v-bind` value is `true`',
```

File: lib/rules/prefer-true-attribute-shorthand.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+const utils = require('../utils')
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    type: 'problem',
```
Comment (FloEdelmann) [inline lib/rules/prefer-true-attribute-shorthand.js:19]: It's only a stylistic/conventional choice, not a problem.

```suggestion
    type: 'suggestion',
```

File: lib/rules/prefer-true-attribute-shorthand.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      report:
+        "Boolean prop with 'true' value can be written in shorthand form.",
+      fix: 'Rewrite this prop into shorthand form.'
+    }
+  },
+  /** @param {RuleContext} context */
+  create(context) {
+    // ...
+
+    return utils.defineTemplateBodyVisitor(context, {
```
Comment (FloEdelmann) [inline lib/rules/prefer-true-attribute-shorthand.js:39]: Please remove the boilerplate comment and empty line:

```suggestion
    return utils.defineTemplateBodyVisitor(context, {
```

File: docs/rules/prefer-true-attribute-shorthand.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+```
+
+```vue
+<template>
+  <MyComponent prop />
+  <MyComponent :prop="true"/>
+</template>
+```
+
+Those two calls will introduce different render result. See [this demo](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCBNeUNvbXBvbmVudCBmcm9tICcuL015Q29tcG9uZW50LnZ1ZSdcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxNeUNvbXBvbmVudCBhIGIvPlxuICA8TXlDb21wb25lbnQgOmE9XCJ0cnVlXCIgOmI9XCJ0cnVlXCIvPlxuPC90ZW1wbGF0ZT4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvdnVlLnJ1bnRpbWUuZXNtLWJyb3dzZXIuanNcIlxuICB9XG59IiwiTXlDb21wb25lbnQudnVlIjoiPHNjcmlwdD5cbmV4cG9ydCBkZWZhdWx0IHtcbiAgcHJvcHM6IHtcbiAgICBhOiBCb29sZWFuLFxuICAgIGI6IFtTdHJpbmcsIEJvb2xlYW5dXG4gIH1cbn1cbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxwcmU+XG5hOiB7e2F9fVxuYjoge3tifX1cbiAgPC9wcmU+XG48L3RlbXBsYXRlPiJ9).
```
Comment (FloEdelmann) [inline docs/rules/prefer-true-attribute-shorthand.md:59]: The order matters, see https://github.com/vuejs/eslint-plugin-vue/issues/1781#issuecomment-1022735437.

In this suggestion, I adapted the explanation and example (also the link) and added a warning box:

````suggestion
:::warning Warning
The shorthand form is not always equivalent! If a prop accepts multiple types, but Boolean is not the first one, a shorthand prop won't pass `true`.
:::

```vue
<script>
export default {
  name: 'MyComponent',
  props: {
    bool: Boolean,
    boolOrString: [Boolean, String],
    stringOrBool: [String, Boolean],
  }
}
</script>
```

**Shorthand form:**

```vue
<MyComponent bool bool-or-string string-or-bool />
```

```
bool: true (boolean)
boolOrString: true (boolean)
stringOrBool: "" (string)
```

**Longhand form:**

```vue
<MyComponent :bool="true" :bool-or-string="true" :string-or-bool="true" />
```

```
bool: true (boolean)
boolOrString: true (boolean)
stringOrBool: true (boolean)
```

Those two calls will introduce different render result. See [this demo](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCBNeUNvbXBvbmVudCBmcm9tICcuL015Q29tcG9uZW50LnZ1ZSdcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIFNob3J0aGFuZCBmb3JtOlxuICA8TXlDb21wb25lbnQgYm9vbCBib29sLW9yLXN0cmluZyBzdHJpbmctb3ItYm9vbCAvPlxuICBcbiAgTG9uZ2hhbmQgZm9ybTpcbiAgPE15Q29tcG9uZW50IDpib29sPVwidHJ1ZVwiIDpib29sLW9yLXN0cmluZz1cInRydWVcIiA6c3RyaW5nLW9yLWJvb2w9XCJ0cnVlXCIgLz5cbjwvdGVtcGxhdGU+IiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwidnVlXCI6IFwiaHR0cHM6Ly9zZmMudnVlanMub3JnL3Z1ZS5ydW50aW1lLmVzbS1icm93c2VyLmpzXCJcbiAgfVxufSIsIk15Q29tcG9uZW50LnZ1ZSI6IjxzY3JpcHQ+XG5leHBvcnQgZGVmYXVsdCB7XG4gIHByb3BzOiB7XG4gICAgYm9vbDogQm9vbGVhbixcbiAgICBib29sT3JTdHJpbmc6IFtCb29sZWFuLCBTdHJpbmddLFxuICAgIHN0cmluZ09yQm9vbDogW1N0cmluZywgQm9vbGVhbl0sXG4gIH1cbn1cbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxwcmU+XG5ib29sOiB7e2Jvb2x9fSAoe3sgdHlwZW9mIGJvb2wgfX0pXG5ib29sT3JTdHJpbmc6IHt7Ym9vbE9yU3RyaW5nfX0gKHt7IHR5cGVvZiBib29sT3JTdHJpbmcgfX0pXG5zdHJpbmdPckJvb2w6IHt7c3RyaW5nT3JCb29sfX0gKHt7IHR5cGVvZiBzdHJpbmdPckJvb2wgfX0pXG4gIDwvcHJlPlxuPC90ZW1wbGF0ZT4ifQ==).
````

**Review by FloEdelmann — CHANGES_REQUESTED** [review]
Thanks for implementing this! However, I have a few remarks.

In https://github.com/vuejs/eslint-plugin-vue/issues/775#issuecomment-454081818, it was suggested to make it configurable whether to always or never use shorthand props. I think that's a good idea, given some people may prefer to never use shorthand props because of their quirks or just by taste.

The `never` case could also be auto-fixable, while the suggestion you implemented would be fine for the `always` case.

**Review by FloEdelmann — CHANGES_REQUESTED** [review]
Nice, this looks really good!

Only one problem I just noticed: The `never` option will also report native HTML shorthand attributes: `<input checked>` should become `<input checked="checked">`, not `<input :checked="true">`. Or is this actually allowed?

If it's allowed, please also add this test case to the "allowed" list to document this behavior.

If it isn't, we need a little more logic: Since one can use native HTML shorthand attributes like `hidden` also on components (e.g. `<MyComponent hidden>`), I think we also have to use suggestions instead of an autofix there: One for changing it into a longhand Vue prop, one for changing it into a longhand HTML attribute.

We might be a bit more clever about this and actually do autofix it in one case though: The shorthand attribute is on a known HTML element. Then we know that it should be autofixed to a longhand HTML attribute. But I think this is a nice-to-have improvement; only having suggestions in all cases would be fine for me.

**Comment by FloEdelmann** [comment]
See also https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html

**Comment by g-plane** [comment]
What about skipping native HTML elements, and just check components?

**Comment by FloEdelmann** [comment]
That'd be fine for me, but still not catch the `<MyComponent hidden>` case. `hidden` might be a prop or an HTML attribute.

**Review by FloEdelmann — APPROVED** [review]
Nice, thanks for the improvements and the test cases! Please run `npm run update` one more time to update the docs (there is no autofix anymore). But this is not merge-blocking; we can do it afterwards.

**Review by FloEdelmann — APPROVED** [review]

**Pull request merged**